### ISSUE TITLE
`Pkg.dev` chooses a registry inconsistently with `Pkg.add`

### ISSUE THREAD

**Issue by ericphanson** [JuliaLang/Pkg.jl#2525]
If you have a package registered with the same name and UUID in two registries, `Pkg.add(name=...)` will choose whichever registry has a later version. However, `Pkg.dev` will choose some other way (whatever is first in some list it checks, I think). Instead, `Pkg.dev` should choose the same registry as `Pkg.add` would. This mismatch can be very confusing for users.

**Comment by fredrikekre** [comment]
Could prompt the user in interactive mode and error otherwise.

**Comment by ericphanson** [comment]
I think that would be an improvement over the current situation, but why not just use the same registry resolution as `add`? Since if you aren’t passing a path or url it needs to use a registry so why not make that consistent.

**Comment by fredrikekre** [comment]
But `Pkg.add` doesn't really choose a registry, it just collapses them all.

**Comment by ericphanson** [comment]
Ah ok, I didn’t realize that. Then I guess it’s a bit harder to implement my suggestion. As a user it still seems like the consistent way for it to behave though.

**Comment by ericphanson** [comment]
But I think a prompt would be a big improvement already. Currently when users dev and get the old version of a package it can make them think something is configured wrong on their end or they have some internal state that is out of date (especially because configuring multiple registries and auth for private registries is not always perfectly smooth). So even if it seems like Pkg should know which registry is the right one, at least giving the options is a big improvement over silently giving the wrong one.

**Comment by fredrikekre** [comment]
With multiple repo urls there is not really a good way to find the canonical one though.

**Comment by ericphanson** [comment]
Well I'm saying there is: in the scenario Pkg is given a name or name + uuid and no URL, then it is hopeless to find a URL unless the package is registered in a registry, in which case it has version numbers in that registry. The case of multiple URLs is when it is registered in multiple registries. Which is the canonical one? The registry which contains the highest version number.

**Comment by fredrikekre** [comment]
> Which is the canonical one? The registry which contains the highest version number.

Why? This seems almost as arbitrary as the current behavior. For example, you have have a private repo and private registry. You decide to open start making releases to a public registry instead, but you probably still want `Pkg.dev` to track your private repo.

Prompting seems simple enough. `Pkg.dev` is inheritly an interactive function so erroring in the non-interactive case seems good anyways.

**Comment by ericphanson** [comment]
> Why?

Since when you `dev`, you get the latest copy of the code. Where is the latest copy of the code likely to be? The one with the latest release.

> For example, you have have a private repo and private registry. You decide to open start making releases to a public registry instead, but you probably still want Pkg.dev to track your private repo.

How are you making releases to a public registry from a private repo?

**Comment by fredrikekre** [comment]
> How are you making releases to a public registry from a private repo?

I am not, I have a public repo with "outdated" code where all releases are from.

**Comment by ericphanson** [comment]
If all the releases are from a public repo, how does Pkg know about the private one at all?

**Comment by ericphanson** [comment]
I think the scenario must be you have registered the package in a public and private registry, then stopped making releases in either but have continued development in the private registry and happen to have the last release be in the public registry.

That just seems way more unlikely than you’re continuing development in whatever repo you’ve last released from.

**Comment by KristofferC** [comment]
So you have two different `repo` URLs in the two different registries. And you say that Pkg picks the "wrong" URL?

**Comment by ericphanson** [comment]
Yep, exactly. This comes about by taking a private package and open sourcing it in a new public repo (in order to not transfer the history to the public one while keeping it around in case some historical git-diving is ever needed) as we did at the [2020 Contributhon](https://julialang.org/blog/2021/01/contributhon/).

So you have the same name and UUID semantically corresponding to the same package that has some versions in a private registry with a URL corresponding to a private repo, and General with a URL corresponding to a public repo with later versions and all future development. `Pkg.add` works great with this process and selects the latest available version, which in this case is in General.

`Pkg.dev` chooses some other way (seemingly arbitrarily) and often chooses the wrong URL to clone.

My claim is that even if this is not the situation for all users of multiple registries who are add'ing and dev'ing packages, having the repo you get from `Pkg.dev` correspond to the repo containing the code you get from `Pkg.add` is consistent behavior, and having `Pkg.dev` choose arbitrarily is not.

**Comment by fredrikekre** [comment]
> If all the releases are from a public repo, how does Pkg know about the private one at all?

Because if you have multiple different URLs there must be more registries.

> I think the scenario must be you have registered the package in a public and private registry, then stopped making releases in either but have continued development in the private repo and happen to have the last release be in the public registry.
>
> That just seems way more unlikely than you’re continuing development in whatever repo you’ve last released from.

Okay, but in that case, why not update the internal repo url? Or are you saying all releases isn't public and Pkg handles that?

**Comment by ericphanson** [comment]
> Because if you have multiple different URLs there must be more registries.

Right, that’s why I don’t understand the scenario you outlined in https://github.com/JuliaLang/Pkg.jl/issues/2525#issuecomment-827658782

> Okay, but in that case, why not update the internal repo url?

Because then Pkg can’t find the old versions because the git history was wiped moving to  the new repo.

> Or are you saying all releases isn't public and Pkg handles that?

No, there’s n private releases (with valid git tree SHAs in the private repo) followed by m public releases (with valid git tree SHAs in the public repo).

**Comment by ericphanson** [comment]
Is it clear what the issue is now?

**Comment by ericphanson** [comment]
I've retitled the issue because the same thing happens with `add MyPkg#branch` as with `dev`. Hopefully this also shows the inconsistency more directly.
